Source/WebCore/html/ValidatedFormListedElement.cpp

```diff
@@ -0,0 +1,494 @@
+    if (m_disabled != newDisabled) {
+        Style::PseudoClassChangeInvalidation disabledInvalidation(asHTMLElement(), { { CSSSelector::PseudoClassDisabled, newDisabled }, { CSSSelector::PseudoClassEnabled, !newDisabled } });
+        m_disabled = newDisabled;
+        disabledAttributeChanged();
+    }
+}
+
+void ValidatedFormListedElement::parseReadonlyAttribute(const AtomString& value)
+{
+    bool newHasReadOnlyAttribute = !value.isNull();
+    if (m_hasReadOnlyAttribute != newHasReadOnlyAttribute) {
+        bool newMatchesReadWrite = supportsReadOnly() && !newHasReadOnlyAttribute;
+        Style::PseudoClassChangeInvalidation readWriteInvalidation(asHTMLElement(), { { CSSSelector::PseudoClassReadWrite, newMatchesReadWrite }, { CSSSelector::PseudoClassReadOnly, !newMatchesReadWrite } });
+        m_hasReadOnlyAttribute = newHasReadOnlyAttribute;
+        readOnlyStateChanged();
+    }
+}
+
+void ValidatedFormListedElement::disabledAttributeChanged()
+{
+    disabledStateChanged();
+}
+
+void ValidatedFormListedElement::insertedIntoAncestor(Node::InsertionType insertionType, ContainerNode& parentOfInsertedTree)
+{
+    resetDataListAncestorState();
+    updateWillValidateAndValidity();
+    syncWithFieldsetAncestors(&parentOfInsertedTree);
+
+    FormListedElement::elementInsertedIntoAncestor(asHTMLElement(), insertionType);
+}
+
+void ValidatedFormListedElement::resetDataListAncestorState()
+{
+    m_dataListAncestorState = DataListAncestorState::Unknown;
+}
+
+void ValidatedFormListedElement::syncWithFieldsetAncestors(ContainerNode* insertionPoint)
+{
+    if (matchesInvalidPseudoClass())
+        addInvalidElementToAncestorFromInsertionPoint(asHTMLElement(), insertionPoint);
+    if (asHTMLElement().document().hasDisabledFieldsetElement())
+        setAncestorDisabled(computeIsDisabledByFieldsetAncestor());
+}
+
+void ValidatedFormListedElement::removedFromAncestor(Node::RemovalType removalType, ContainerNode& oldParentOfRemovedTree)
+{
+    HTMLElement& element = asHTMLElement();
+    bool wasMatchingInvalidPseudoClass = matchesInvalidPseudoClass();
+
+    m_validationMessage = nullptr;
+    if (m_disabledByAncestorFieldset)
+        setAncestorDisabled(computeIsDisabledByFieldsetAncestor());
+
+    bool wasInsideDataList = false;
+    if (m_dataListAncestorState == DataListAncestorState::InsideDataList) {
+        m_dataListAncestorState = DataListAncestorState::Unknown;
+        wasInsideDataList = true;
+    }
+
+    FormListedElement::elementRemovedFromAncestor(element, removalType);
+
+    if (wasMatchingInvalidPseudoClass)
+        removeInvalidElementToAncestorFromInsertionPoint(element, &oldParentOfRemovedTree);
+
+    if (wasInsideDataList)
+        updateWillValidateAndValidity();
+}
+
+bool ValidatedFormListedElement::computeIsDisabledByFieldsetAncestor() const
+{
+    RefPtr<Element> previousAncestor;
+    for (RefPtr<Element> ancestor = asHTMLElement().parentElement(); ancestor; ancestor = ancestor->parentElement()) {
```

Comment:
Use lineageOfType<Element>(asHTMLElement()) (include ElementAncestorIterator.h).

Reply:
`ancestorsOfType(asHTMLElement()` actually haha (learned the hard way), fixed in [451fc8e](https://github.com/WebKit/WebKit/pull/4988/commits/451fc8e4bc29c8669d21b87393084d5082dd88ac).